File: src/test/java/com/bso/dracko/mediator/springboot/SpringRegistryTest.java
package com.bso.dracko.mediator.springboot;

import com.bso.dracko.mediator.contract.*;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest(classes = SpringBootTestConfiguration.class, webEnvironment = SpringBootTest.WebEnvironment.NONE)
@Import({ SpringRegistryTest.SpringRegistryTestConfig.class })
class SpringRegistryTest {

    @Autowired
    private ApplicationContext applicationContext;

    @Autowired
    private Registry registry;

    @Configuration
    public static class SpringRegistryTestConfig {
        @Bean
        public CommandHandler<DummyCommand> commandHandler() {
            return new DummyCommandHandler();
        }
        @Bean
        public RequestHandler<DummyRequest, String> requestHandler() {
            return new DummyRequestHandler();
        }
        @Bean
        public EventHandler<DummyEvent> eventHandler() {
            return new DummyEventHandler();
        }
    }

    private static class DummyCommand implements Command {}
    private static class DummyRequest implements Request<String> {}
    private static class DummyEvent implements Event {}

    private static class DummyRequestHandler implements RequestHandler<DummyRequest, String> {
        @Override
        public String handle(DummyRequest dummyRequest) { return "xpto"; }
    }
    private static class DummyCommandHandler implements CommandHandler<DummyCommand> {
        @Override
        public void handle(DummyCommand dummyCommand) { /* do nothing */ }
    }
    private static class DummyEventHandler implements EventHandler<DummyEvent> {
        @Override
        public void handle(Event event) { /* do nothing */ }
    }

    @Test
    void contextLoadRegistryBeanTest() {
        Registry registryBean = applicationContext.getBean(Registry.class);
        assertNotNull(registryBean);
        assertTrue(registryBean instanceof SpringRegistry);
    }

    @Test
    void getCommandHandlerTest() {
        DummyCommand cmd = new DummyCommand();
        CommandHandler<DummyCommand> commandHandler = registry.getCommandHandler(cmd);
        assertNotNull(commandHandler);
    }

    @Test
    void getEventHandlerTest() {
        var event = new DummyEvent();
        List<EventHandler<DummyEvent>> eventHandlers = registry.getEventHandler(event);
        assertNotNull(eventHandlers);
        assertEquals(1, eventHandlers.size());
    }

    @Test
    void getRequestHandlerTest() {
        Request<String> req = new DummyRequest();
        RequestHandler<Request<String>, String> requestHandler = registry.getRequestHandler(req);
        assertNotNull(requestHandler);
        String result = requestHandler.handle(req);
        assertEquals("xpto", result);
    }
}

File: src/test/kotlin/com/bso/dracko/mediator/springboot/SpringRegistryTest.kt
package com.bso.dracko.mediator.springboot

import com.bso.dracko.mediator.contract.Command
import com.bso.dracko.mediator.contract.CommandHandler
import com.bso.dracko.mediator.contract.Event
import com.bso.dracko.mediator.contract.EventHandler
import com.bso.dracko.mediator.contract.Registry
import com.bso.dracko.mediator.contract.Request
import com.bso.dracko.mediator.contract.RequestHandler
import com.bso.dracko.mediator.springboot.config.SpringBootIntegrationTestConfiguration
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import

@SpringBootTest(classes = [SpringBootIntegrationTestConfiguration::class], webEnvironment = SpringBootTest.WebEnvironment.NONE)
@Import(SpringRegistryTest.SpringRegistryTestConfig::class)
class SpringRegistryTest {

    @Autowired
    private lateinit var applicationContext: ApplicationContext

    @Autowired
    private lateinit var registry: Registry

    class DummyCommand : Command
    class DummyRequest : Request<String>
    class DummyEvent : Event

    class DummyCommandHandler : CommandHandler<DummyCommand> {
        override fun handle(command: DummyCommand) { }
        override fun getType(): Class<DummyCommand> = DummyCommand::class.java
    }

    class DummyRequestHandler : RequestHandler<DummyRequest, String> {
        override fun handle(request: DummyRequest) = "xpto"
        override fun getType(): Class<DummyRequest> = DummyRequest::class.java
    }

    class DummyEventHandler : EventHandler<DummyEvent> {
        override fun handle(event: DummyEvent) { }
        override fun getType(): Class<DummyEvent> = DummyEvent::class.java
    }

    @Configuration
    class SpringRegistryTestConfig {
        @Bean
        fun commandHandler(): CommandHandler<DummyCommand> {
            return DummyCommandHandler()
        }

        @Bean
        fun requestHandler(): RequestHandler<DummyRequest, String> {
            return DummyRequestHandler()
        }

        @Bean
        fun eventHandler(): EventHandler<DummyEvent> {
            return DummyEventHandler()
        }
    }

    @Test
    fun contextLoadRegistryBeanTest() {
        val registryBean = applicationContext.getBean(Registry::class.java)
        Assertions.assertNotNull(registryBean)
        Assertions.assertTrue(registryBean is SpringRegistry)
    }

    @Test
    fun getCommandHandlerTest() {
        val cmd = DummyCommand()
        val commandHandler = registry.getCommandHandler(cmd)
        Assertions.assertNotNull(commandHandler)
    }

    @Test
    fun getEventHandlerTest() {
        val event = DummyEvent()
        val eventHandlers = registry.getEventHandler(event)
        Assertions.assertNotNull(eventHandlers)
        Assertions.assertEquals(1, eventHandlers.size)
    }

    @Test
    fun getRequestHandlerTest() {
        val req: Request<String> = DummyRequest()
        val requestHandler = registry.getRequestHandler(req)
        Assertions.assertNotNull(requestHandler)
        val result = requestHandler!!.handle(req)
        Assertions.assertEquals("xpto", result)
    }
}
